File: src/init/load-input-tracker.cpp
#include "init/load-input-tracker.hpp"

#include <fstream>
#include <memory>
#include <unordered_map>
#include "engine/input-system/include.hpp"
#include "engine/resource-system/json/include.hpp"
#include "RawInputSFML.hpp"

engine::inputsystem::InputTracker loadInputTracker(const std::string& controlsFileName) {
    using namespace engine::inputsystem;
    std::ifstream controlsFile(controlsFileName);

    auto keyMappingJson = parseJSON(controlsFile);
    std::unordered_map<KeyboardKey, GameKey> keyMapping;

    for (const auto& [keyboardKey, gameKey] : keyMappingJson.asIterableMap()) {
        keyMapping.insert({
            keyFromString(keyboardKey),
            gameKey.asString()
        });
    }

    return InputTracker(std::make_unique<RawInputSFML>(), keyMapping);
}

File: frontend/src/init/load-input-tracker.ts
import { InputTracker } from '../engine/input-system/InputTracker';
import { KeyboardKey } from '../engine/input-system/KeyboadKey';
import { GameKey } from '../engine/input-system/types';
import { RawInputBrowser } from '../RawInputBrowser';

interface ControlsMap {
    [keyboardKey: string]: string;
}

export function loadInputTracker(controlsMap: ControlsMap): InputTracker {
    const rawInput = new RawInputBrowser();
    rawInput.setupKeyWatcher();
    const keyMapping = buildKeyMapping(controlsMap);
    return new InputTracker(rawInput, keyMapping);
}

function buildKeyMapping(controlsMap: ControlsMap): Map<KeyboardKey, GameKey> {
    const keyMapping = new Map<KeyboardKey, GameKey>();

    for (const keyboardKey in controlsMap) {
        const gameKey = controlsMap[keyboardKey];
        keyMapping.set(KeyboardKey[keyboardKey as keyof typeof KeyboardKey], gameKey);
    }

    return keyMapping;
}

